### PR TITLE
fix Julia 1.7 tests

### PR DESCRIPTION
**Pull request by st--**

### CODE REVIEW

**Comment by willtebbutt** [comment]
I'm a little bit confused as to why this improves type-stability. Do we have a global `cache` variable somewhere or something?

**Comment by st--** [comment]
Actually, I think the `something()` in the return line makes it type-stable already, the `cache` variable is only locally type-unstable (initialized as `nothing`, then overwritten with something). But I still think it's cleaner with `nonlocal`.

**Comment by st--** [comment]
Test failure is due to optimisation having ended up in a bad local optimum ...

**Comment by willtebbutt** [comment]
> Test failure is due to optimisation having ended up in a bad local optimum ...

Hmmm that's weird. Shouldn't the optimisation problem be convex, since the likelihood is log-concave?

**Comment by st--** [comment]
> > Test failure is due to optimisation having ended up in a bad local optimum ...
> 
> Hmmm that's weird. Shouldn't the optimisation problem be convex, since the likelihood is log-concave?

The inner loop should be convex optimisation, but not the optimisation of hyperparameters using the approximate log marginal likelihood..

But yeah I don't understand why that would have changed so much in the 1.6 -> 1.7 change :S

**Comment by rossviljoen** [comment]
Might be because they changed the default RNG?

**Comment by willtebbutt** [comment]
I'm still a bit confused by the need for `local`. AFAICT, there's not a `cache` variable in the module-level scope of `ApproximateGPs`, so is the `local` keyword no a no-op? How would what we see differ if `local` were not used?

File: test/laplace.jl [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
         X = range(0, 23.5; length=48)
-        fs = @. 3 * sin(10 + 0.6X) + sin(0.1X) - 1
-        # invlink = normcdf
-        invlink = logistic
-        ps = invlink.(fs)
-        Y = [rand(Bernoulli(p)) for p in ps]
+        # The random number generator changed in 1.6->1.7. The following vector was generated in Julia 1.6.
+        # The generating code below is only kept for illustrative purposes.
+        #! format: off
+        Y = [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0]
```
Comment (rossviljoen) [inline test/laplace.jl:7]: Couldn't we just use `rng=MersenneTwister()` or whatever?

**Comment by rossviljoen** [comment]
@willtebbutt It's just a nicer way of initialising the variable no?

**Review by rossviljoen — APPROVED** [review]

**Comment by willtebbutt** [comment]
> @willtebbutt It's just a nicer way of initialising the variable no?

🤦 you're right.

**Review by willtebbutt — APPROVED** [review]
LGTM

**Pull request merged**